modules/querier/querier.go

```diff
@@ -326,23 +324,23 @@ func (q *Querier) FindTraceByID(ctx context.Context, req *tempopb.TraceByIDReque
 }
 
 type (
-	forEachFn        func(ctx context.Context, client tempopb.QuerierClient) (interface{}, error)
+	forEachFn        func(ctx context.Context, client tempopb.QuerierClient) error
 	replicationSetFn func(r ring.ReadRing) (ring.ReplicationSet, error)
 )
 
 // forIngesterRings runs f, in parallel, for given ingesters
-func (q *Querier) forIngesterRings(ctx context.Context, userID string, getReplicationSet replicationSetFn, f forEachFn) ([]responseFromIngesters, error) {
+func (q *Querier) forIngesterRings(ctx context.Context, userID string, getReplicationSet replicationSetFn, f forEachFn) error {
```

Comment:
I really like the new approach of just a simple callback.  It avoids both interface typecasting, even the original thought of making it use generics. 👍

Reply:
we can make the similar change in `forGivenGenerators` https://github.com/grafana/tempo/blob/807435d38ac5b78c5626d05d74eb0715081f1066/modules/querier/querier.go#L432 but I left it out to keep this PR focused on ingester changes. 

can make a follow up PR, if you want to make that change :)

Reply:
pushed [`e1139f4` (#4119)](https://github.com/grafana/tempo/pull/4119/commits/e1139f4e2200d9353615b58ee33361bab0c64fd8)

PR is in draft because it's branched from this PR's branch.